### ISSUE TITLE
has(key) does not use index

### ISSUE THREAD

**Issue by s-nel** [JanusGraph/janusgraph#284]
The Tinkerpop docs say:

> `has(key)`: Remove the traverser if its element does not have a value for the key.

I would expect the following query to use an index I've created on property key "foo" and label "bar"

```
g.V().has(Key[String]("foo")).hasLabel("bar")
```

but it instead results in a full graph scan. Here's a minimal example (Scala):

```
    val conf = new BaseConfiguration()
    conf.setProperty("storage.backend","inmemory")
    val graph = JanusGraphFactory.open(conf)

    val mgmt = graph.openManagement()
    val nameKey = mgmt.getOrCreatePropertyKey("name")
    val planetLabel = mgmt.getOrCreateVertexLabel("planet")
    mgmt.buildIndex("planetByName", classOf[Vertex]).addKey(nameKey).indexOnly(planetLabel).buildCompositeIndex()
    mgmt.commit()

    (1 to 5) foreach { i ⇒
      graph.addVertex(T.label, "planet", "name", s"planet $i")
    }

    println("Fetching planets")
    graph.traversal().V().hasLabel("planet").count().next() shouldBe 5 // Expected to be a full graph scan
    println("Fetching planet 1")
    graph.traversal().V().has("name", "planet 1").hasLabel("planet").count().next() shouldBe 1 // Expected to be an index hit
    println("Fetching planets with a value for name")
    graph.traversal().V().has("name").hasLabel("planet").count().next() shouldBe 5 // Expected to be an index hit, but is a full graph scan!!

    graph.close
```

Janus 0.1.1

**Comment by davidclement90** [comment]
Hi,
Your index is define only on vertex "planet" and your request say that you search on vertex "bar".
You have two solutions : 
- you define your index on all vertices labels.
- you define a second index on vertex "planet".

David.

**Comment by s-nel** [comment]
"foo" and "bar" were for example. As you can see in my test, I create an index on name and planet label

```
mgmt.buildIndex("planetByName", classOf[Vertex]).addKey(nameKey).indexOnly(planetLabel).buildCompositeIndex()
```

then when i query it later 

```
graph.traversal().V().has("name").hasLabel("planet").count().next()
```

it's a full graph scan.

**Comment by davidclement90** [comment]
Your are right "has" do not use composite index, I read your issue to fast.
I think it is because the possibility to make an "exists" query was introduce in ES2. JanusGraph support ES2 only in master.
I think it would be a good improvement.
Until this feature, has("name", neq("")) should work in most cases.

**Comment by laughedelic** [comment]
> Until this feature, you can try has("name", neq("")).

I think it's not a workaround. I'm using somewhat similar:

```
graph.query().hasNot("name", "").vertices
```

and it gives me warnings about iterating over all vertices.

**Comment by ankurgoelmb** [comment]
Any plan when this improvement will be available. This is important for graphDB, where all vertex can not have symmetric properties, and to create relation between two vertex based on same property  index on property is must.

**Comment by HackingJ** [comment]
I think that you need to below api.

`has(final String label, final String propertyKey, final Object value)`

**Comment by m-manseer** [comment]
Has this issue been fixed? Seems like I've run into the same problem with 0.2.1. I want to find all vertices that have a certain property. There's a composite index on that property installed (on all labels).

Using `.hasNot(..., "")` or `.has(..., P.neq(""))` does not work either.

**Comment by sunsided** [comment]
It's still a problem on JanusGraph 0.4.0.

**Comment by li-boxuan** [comment]
To my best knowledge, the problem "has(key) does not use COMPOSITE index" is due to the composite index design and thus I don't think this could be "fixed" with keeping backward-compatibility.

In JanusGraph, a composite index entry uses (index name + value) as the key (might be hashed) in the index store. I don't see a way to query just by index name unless we change the way how the composite index is stored.

The workaround can be to use a mixed index or to create another composite index B which records whether a record has composite index A or not.

**Comment by li-boxuan** [comment]
To my best knowledge, this is not possible for a composite index (at least for Cassandra). JanusGraph uses a combination of index name and index value (might also be hashed) as the partition key (again, in Cassandra). If you want to query by index name only, depending on your partition strategy, it might require a full scan of the index store. Depending on the nature of your data and indexes, using the composite index, in this case, might make your performance worse.

**Comment by FlorianHockmann** [comment]
@li-boxuan implemented support for this in #2175 for mixed indices and it seems like there isn't a good way to support this based on a composite index.
So, shouldn't we close this issue now? Or should we document how `.has(key)` can use a mixed index?

What do you think @li-boxuan?

**Comment by li-boxuan** [comment]
I'd say let's close the issue. I didn't add to user doc because this is internal optimization that users shouldn't need to know about.
